Server Access RBAC Docs page (#20246)
gravitational/teleport: docs/pages/server-access/rbac.mdx

---
title: Access Controls for Servers
description: Role-based access control (RBAC) for Teleport Server Access.
---

> # Server Access Role-Based Access Control

Reviewer: ```suggestion
```
As suggested in my last review

You can use Teleport's role-based access control (RBAC) system to set up
granular permissions for authenticating to Linux servers connected to Teleport.

An example of a policy could be, *Server administrators have access to
everything, the QA team and engineers have full access to staging servers, and
engineers can gain temporary access to production servers in case of

> emergency".

Reviewer: ```suggestion
emergency.*
```
To balance the italics markup

For a more general description of Teleport roles and examples see [RBAC](../access-controls/introduction.mdx), as
this section focuses on configuring RBAC for Server Access.

## Role configuration

Teleport's "role" resource provides the following instruments for restricting
server access:

```yaml
kind: role
version: v5
metadata:
  name: developer
spec:
  allow:
    # The logins array defines the OS/UNIX logins a user is allowed to use.
    # both strings and template variables are supported in this field
    logins: [ubuntu, debian, '{{internal.logins}}']

    # node_labels: a user with this role will be allowed to connect to
    # SSH nodes whose labels match any of the expressions below.
    node_labels:
      # literal strings:
      'env': 'test'
      # the wildcard ('*') means "any node"
      '*': '*'
      # a list of allowed values:
      'region': ['us-west-1', 'eu-central-1']
      # regular expressions start with ^ and end with $

>       # Teleport uses golang regexp syntax.

Reviewer: ```suggestion
      # Teleport uses Golang's regexp syntax.
```

      # the list example above can be expressed as:
      'reg': '^us-west-1|eu-central-1$'

    # List of host groups the created user will be added to. Any that don't already exist are created.
    host_groups: [ubuntu, nginx, other]

    # Assign the user to the sudoers group
    host_sudoers:
    - 'ALL=(ALL) NOPASSWD: ALL'

    # Deny access to the root user
  deny:
    logins:
    - root
```

<Admonition
  type="note"
  title="Deny Rules"
>
  Deny rules will match greedily. In the example above, a server session
  attempting to use "root" serveruser  account will be rejected.
</Admonition>

## Template variables

> Similar to other role fields, the Server related fields support templating variables.

Reviewer: ```suggestion
Similar to role fields for accessing other resources in Teleport, server-related fields support template variables.
```
To make the first part of the sentence more explicit

> The `{{external.xyz}}` variables are replaced with values from external [SSO](../access-controls/sso.mdx)

Reviewer: ```suggestion
Variables with the format `{{external.xyz}}` are replaced with values from external [SSO](../access-controls/sso.mdx)
```

providers. For OIDC logins, `{{external.xyz}}` refers to the "xyz" claim; for
SAML logins, `{{external.xyz}}` refers to the "xyz" assertion.

For example, here is what a role may look like if you want to assign allowed
server environment types and allowed logins from the user's Okta `environments` and 
`allowedlogins` assertions:

```yaml
spec:
  allow:
    node_labels:
    - env:  '{{external.environments}}'
    logins:
    - '{{external.allowedlogins}}'
```

The `{{internal.logins}}` variable permits sharing
allowed server logins with remote clusters. They will be replaced
with the respective properties of a remote user connecting from a root cluster.

> For example, suppose a user in the root cluster has the following role:

Reviewer: We haven't introduced the term "root cluster" yet on this page. I think it would make sense to mention Trusted Clusters here if we're going to use the term "root cluster". The `internal` prefix also works without Trusted Clusters, though, right?

```yaml
spec:
  allow:
    logins: ["ubuntu"]
```

> The role on the leaf cluster can be set up to use the user's allowed server

Reviewer: Same as above re: Trusted Clusters

Author: Updated

accounts and names:

```yaml
spec:
  allow:
    logins: ["{{internal.logins}}"]
```

## Server role options

The `allow` and `deny` sections described above are used for controlling
the servers and logins allowed. Role options provide the Teleport features
available for users with the specified role. These options apply to Server access.

```yaml
spec:
  allow:
    #....
  options:
    # Controls whether this role supports auto provisioning of users.
    create_host_user: true
    # forward_agent controls whether SSH agent forwarding is allowed
    forward_agent: true
    # port_forwarding controls whether TCP port forwarding is allowed for SSH
    port_forwarding: true
    # ssh_file_copy controls whether file copying (SCP/SFTP) is allowed.
    # Defaults to true.
    ssh_file_copy: false
    # client_idle_timeout determines if SSH sessions to cluster nodes are
    # forcefully terminated after no activity from a client (idle client).

>     # it overrides the global cluster setting. examples: "30m", "1h" or "1h30m"

Reviewer: ```suggestion
    # It overrides the global cluster setting. examples: "30m", "1h" or "1h30m"
```

    client_idle_timeout: never
    # Determines if the clients will be forcefully disconnected when their
    # certificates expire in the middle of an active session.
    # It overrides the global cluster setting.
    disconnect_expired_cert: no
    # max_sessions is total number of session channels that can be established
    # across a single connection. Setting it to 10 matches OpenSSH default behavior.
    max_sessions: 10
    # Defines which events are recorded by the BPF-based session recorder.
    enhanced_recording:
    - command
    - disk
    - network
    # permit_x11_forwarding allows users to use X11 forwarding with openssh
    # clients and servers through the proxy
    permit_x11_forwarding: true

>     # enterprise-only max_connections field sets a limit of concurrent sessions within a

Reviewer: ```suggestion
    # The Enterprise-only max_connections field sets a limit of concurrent sessions within a
```

    # cluster. This setting slows down Teleport performance because it has to track
    # connections cluster-wide.
    max_connections: 2
    # Define how Teleport deals with session recording failures, such as a full
    # disk error. The value can be set to either `best_effort` or `strict`. If
    # set to `strict`, the session will terminate immediately. If set to
    # `best_effort`, the session won’t be terminated, and the recording will be
    # disabled. The configuration is done per service (currently, only `ssh` is
    # supported).
    record_session:
      # Optional: the default session recording mode to use when a
      # protocol-specific mode is not set.
      default: best_effort|strict

>       # Optional: Session recording mode for SSH sessions (Teleport Server
>       # Access). If not set, the value set on default will be used.

Reviewer: ```suggestion
      # Optional: Session recording mode for SSH sessions. 
      # If not set, the value set on default will be used.
```
Since "Server Access" is no longer a product

      ssh: best_effort|strict

>     # enterprise-only: when enabled, the source IP that was used to log in is embedded in the SSH

Reviewer: ```suggestion
    # Enterprise-only: when enabled, the source IP that was used to log in is embedded in the SSH
```

---
title: Server Access Access Controls
description: Role-based access control (RBAC) for Teleport Server Access.
---

# Server Access Role-Based Access Control

Role-based access control (or RBAC, for short) allows administrators to set up
granular access policies for Linux servers connected to Teleport.

An example of a policy could be, *""server administrators have access to
everything, QA team and engineers have full access to staging servers, and
engineers can gain temporary access to the production server in case of
emergency"*.

For a more general description of Teleport roles and examples see [RBAC](../access-controls/introduction.mdx), as
this section focuses on configuring RBAC for Server Access.

## Role configuration

Teleport's "role" resource provides the following instruments for restricting
server access:

```yaml
kind: role
version: v5
metadata:
  name: developer
spec:
  allow:
    # The logins array defines the OS/UNIX logins a user is allowed to use.
    # both strings and template variables are supported in this field
    logins: [ubuntu, debian, '{{internal.logins}}']

    # node_labels: a user with this role will be allowed to connect to

>     # SSH nodes, which labels match expressions below.

Reviewer 2: ```suggestion
    # SSH nodes whose labels match any of the expressions below.
```

    node_labels:
      # literal strings:
      'env': 'test'
      # the wildcard ('*') means "any node"
      '*': '*'

>       # a list of alternative options:

Reviewer 2: ```suggestion
      # a list of allowed values:
```

      'region': ['us-west-1', 'eu-central-1']
      # regular expressions start with ^ and end with $
      # Teleport uses golang regexp syntax.

>       # of the list example above can be expressed as:

Reviewer 2: ```suggestion
      # the list example above can be expressed as:
```

      'reg': '^us-west-1|eu-central-1$'

    # List of host groups the created user will be added to. Any that don't already exist are created.
    host_groups: [ubuntu, nginx, other]

    # Assign the user to the sudoers group
    host_sudoers:
    - 'ALL=(ALL) NOPASSWD: ALL'

    # Deny access to the root user
  deny:
    logins:
    - root

>   options:

Reviewer 2: This section of the page is about restricting access. Role options are covered below.

I would remove options from here to avoid introducing too many concepts at once.

Author: addressed

   # Automatically create the user if does not exist
   # The user is deleted after the session ends if Teleport created the user. 
    create_host_user: true
```

<Admonition
  type="note"
  title="Deny Rules"
>
  Deny rules will match greedily. In the example above, a server session
  attempting to use "root" serveruser  account will be rejected.
</Admonition>

## Template variables

Similar to other role fields, the Server related fields support templating variables.

The `{{external.xyz}}` variables are replaced with values from external [SSO](../access-controls/sso.mdx)
providers. For OIDC, they will be expanded with a value of an "xyz" claim; for

> SAML — with an "xyz" assertion value.

Reviewer 2: For OIDC logins, `{{external.xyz}}` refers to the "xyz" claim.
For SAML logins, `{{external.xyz}}` refers to the "xyz" assertion.

Author: updated phrasing

For example, here is what a role may look like if you want to assign allowed
server environment types from the user's Okta `environments` assertion:

```yaml
spec:
  allow:
    node_labels:
    - env:  '{{external.environments}}'
```

The `{{internal.logins}}` variable permits sharing
allowed server logins with remote clusters. They will be replaced
with the respective properties of a remote user connecting from a root cluster.

For example, suppose a user in the root cluster has the following role:

```yaml
spec:
  allow:
    logins: ["ubuntu"]
```

The role on the leaf cluster can be set up to use the user's allowed server
accounts and names:

```yaml
spec:
  allow:
    logins: ["{{internal.logins}}"]
```

## Server role options

> The following options in a Teleport role apply to Server Access.

Reviewer 2: A one-sentence introduction that describes what a role option is would be helpful here. Maybe something along the lines of:

The allow and deny sections descried above are used for controlling access.
Role options offer the ability to control which Teleport features are available
for users with the specified role.

Author: added longer intro along these lines

> ```yaml

Reviewer 2: Would this be a good candidate for a partial? I assume there's very similar content on the main RBAC page, right?

Author: As there are differences in things like the recorded sessions area I'd say no.  Otherwise we'd have to have multiple partial splits.

    host_sudoers:
    - 'ALL=(ALL) NOPASSWD: ALL'

    # Deny access to the root user
  deny:
    logins:
    - root
```

<Admonition
  type="note"
  title="Deny Rules"
>
  Deny rules will match greedily. In the example above, a server session
  attempting to use the "root" server user  account will be rejected.
</Admonition>

## Template variables

Similar to role fields for accessing other resources in Teleport, server-related fields support template variables.

Variables with the format `{{external.xyz}}` are replaced with values from external [SSO](../access-controls/sso.mdx)
providers. For OIDC logins, `{{external.xyz}}` refers to the "xyz" claim; for
SAML logins, `{{external.xyz}}` refers to the "xyz" assertion.

For example, here is what a role may look like if you want to assign allowed
server environment types and allowed logins from the user's Okta `environments` and 
`allowedlogins` assertions:

```yaml
spec:
  allow:
    node_labels:
    - env:  '{{external.environments}}'
    logins:
    - '{{external.allowedlogins}}'
```

The `{{internal.logins}}` variable applies to local users
and works with Teleport Trusted Clusters.  Trusted Clusters allows

> connecting via a root Teleport Cluster to resources connected to other Teleport Clusters.
> Those Teleport Clusters, identified as leaf clusters, allow the connection
> by trusting the root Teleport Cluster.

Reviewer: ```suggestion
connecting via a root Teleport cluster to resources connected to other Teleport clusters.
Those Teleport clusters, identified as leaf clusters, allow the connection
by trusting the root Teleport cluster.
```
